lnwallet/channel.go

```diff
@@ -215,50 +228,93 @@ type commitment struct {
 // toChannelDelta converts the target commitment into a format suitable to be
 // written to disk after an accepted state transition.
 // TODO(roasbeef): properly fill in refund timeouts
-func (c *commitment) toChannelDelta() (*channeldb.ChannelDelta, error) {
+func (c *commitment) toChannelDelta(ourCommit bool) (*channeldb.ChannelDelta, error) {
 	numHtlcs := len(c.outgoingHTLCs) + len(c.incomingHTLCs)
+
+	// Save output indexes for RHash values found, so we don't return the
+	// same output index more than once.
+	dups := make(map[PaymentHash][]uint16)
 	delta := &channeldb.ChannelDelta{
 		LocalBalance:  c.ourBalance,
 		RemoteBalance: c.theirBalance,
 		UpdateNum:     c.height,
 		Htlcs:         make([]*channeldb.HTLC, 0, numHtlcs),
 	}
 
+	// Check to see if element (e) exists in slice (s)
+	contains := func(s []uint16, e uint16) bool {
+		for _, a := range s {
+			if a == e {
+				return true
+			}
+		}
+		return false
+	}
+
 	// As we also store the output index of the HTLC for continence
 	// purposes, we create a small helper function to locate the output
 	// index of a particular HTLC within the current commitment
 	// transaction.
-	locateOutputIndex := func(p *PaymentDescriptor) uint16 {
+	locateOutputIndex := func(p *PaymentDescriptor) (uint16, error) {
 		var idx uint16
+		var found bool
+
+		pkScript := p.theirPrevPkScript
+		if ourCommit {
+			pkScript = p.ourPkScript
+		}
 		for i, txOut := range c.txn.TxOut {
-			// TODO(roasbeef): duplicated payment hashes...
-			if bytes.Equal(txOut.PkScript, p.pkScript) {
+			if bytes.Equal(txOut.PkScript, pkScript) &&
+				txOut.Value == int64(p.Amount) {
+				if contains(dups[p.RHash], uint16(i)) {
+					continue
+				}
+				found = true
 				idx = uint16(i)
+				dups[p.RHash] = append(dups[p.RHash], idx)
 				break
 			}
 		}
-		return idx
+		if !found {
+			return 0, fmt.Errorf("could not find a matching output for the HTLC " +
+				"in the commitment transaction")
+		}
+		return idx, nil
 	}
 
+	var index uint16
+	var err error
 	for _, htlc := range c.outgoingHTLCs {
+		if (ourCommit && htlc.isDustLocal) || (!ourCommit && htlc.isDustRemote) {
```

Comment:
Minor rewrite to (IMO) make this fragment a bit more readable (and also adhere to the style guide): (breaking at the `||`) 
```go
if (ourCommit && htlc.isDustLocal) || 
   (!ourCommit && htlc.isDustRemote) {
}
```